File: src/main/java/com/baidu/algorithm/greedy/_406_Queue_Reconstruction_by_Height.py
class Solution(object):
    def reconstructQueue(self, people):
        """
        :type people: List[List[int]]
        :rtype: List[List[int]]
        """
        if not people:
            return []

        dt, height, res = {}, [], []
        for p in people:
            if p[0] not in dt:
                # dt[p[0]] = []
                height.append(p[0])

            # dt[p[0]].append(p[1])
            dt.setdefault(p[0], []).append(p[1])

        height.sort(reverse=True)

        for h in height:
            dt[h].sort()
            for p in dt[h]:
                res.insert(p, [h, p])

        return res

File: src/main/java/com/baidu/algorithm/greedy/_406_Queue_Reconstruction_by_Height.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.greedy;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * _406_Queue_Reconstruction_by_Height
 *
 * @author xuhaoran01
 */
public class _406_Queue_Reconstruction_by_Height {
    public int[][] reconstructQueue(int[][] people) {
        if (people == null || people.length == 0) {
            return new int[0][0];
        }

        Arrays.sort(people, (x, y) -> {
            if (x[0] != y[0]) {
                return y[0] - x[0];
            } else {
                return x[1] - y[1];
            }
        });

        List<int[]> res = new ArrayList<>();
        for (int[] p : people) {
            res.add(p[1], p);
        }

        return res.toArray(new int[0][0]);
    }
}
